### PR TITLE
Sprint1

### PR DESCRIPTION
**Pull request by ADDY69O**

### CODE REVIEW

File: src/app/api.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { HttpInterceptorFn, HttpResponse, HttpErrorResponse, HttpEvent, HttpHandler, HttpRequest } from '@angular/common/http';
+import { throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+export interface APIResponse<T> {
```
Comment (thesaty4) [inline src/app/api.interceptor.ts:5]: move to specific type file

File: src/app/api.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  code: number;
+  status: string;
+  message: string;
+  data: T;
+}
+
+export const apiInterceptor: HttpInterceptorFn = (req, next) => {
```
Comment (thesaty4) [inline src/app/api.interceptor.ts:12]: whatever you will, everything you need to know

File: src/app/api.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+  return next(req).pipe(
+    map(event => {
+      if (event instanceof HttpResponse) {
```
Comment (thesaty4) [inline src/app/api.interceptor.ts:17]: read more about it.

File: src/app/api.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      
+
+        // Transforming the response into the APIResponse format
+        const apiResponse: APIResponse<any> = {
```
Comment (thesaty4) [inline src/app/api.interceptor.ts:21]: why any ?

File: src/app/api.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          message: 'Request successful',  // Customize this message as needed
+          data: event.body
+        };
+
+        // Clone the HttpResponse and replace the body with the transformed apiResponse
+        return event.clone({ body: apiResponse });
+      }
+      return event;
+    }),
+    catchError((error: HttpErrorResponse) => {
```
Comment (thesaty4) [inline src/app/api.interceptor.ts:33]: we can handle or customise the error from here.

File: src/app/api.service.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { BaseApiService } from "./base-api.service";
+import { HttpClient } from "@angular/common/http";
+import { CacheService } from "./cache.service";
+
+
+export class ApiService extends BaseApiService {
+
+    constructor ( http: HttpClient, cacheService: CacheService){
+        super(http,`https://dummyjson.com/`,cacheService)
```
Comment (thesaty4) [inline src/app/api.service.ts:9]: endpoint, baseAPI

File: src/app/api.service.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+
+    getPosts <T> (limit:T,skip:T):void{
```
Comment (thesaty4) [inline src/app/api.service.ts:13]: wrong

File: src/app/base-api.service.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import { Injectable } from '@angular/core';
+import { CacheService } from './cache.service';
+
+export interface GetDataType {
+    title:string,
+    body:string,
+  }
+
+@Injectable({
+  providedIn: 'root'
```
Comment (thesaty4) [inline src/app/base-api.service.ts:12]: remove unused file & code

File: src/app/pagination/pagination.component.css [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  h1 {
+    text-align: center;
+    color: #333;
+  }
+
+  .loader-overlay {
+    position: fixed;
+    top: 0;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:14]: in one line

File: src/app/pagination/pagination.component.css [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    border-top: 1px solid #ccc;
+    display: flex;
+    flex-direction: row;
+    flex-wrap: wrap;
+    justify-content: space-around;
+    width: 98%;
+    border: 2px solid #333;
+    overflow-y: scroll;
+    height: 65vh;
+    background-color: #f9f9f9;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:41]: color variables

File: src/app/pagination/pagination.component.css [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    padding: 10px;
+    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
+  }
+
+  .post {
+    width: 30%;
+    margin: 10px 0 0 10px;
+    background-color: #fff;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:49]: same as above

File: src/app/pagination/pagination.component.css [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  .post h2 {
+    color: #007bff;
+    font-size: 20px;
+    margin-bottom: 10px;
+  }
+
+  .post p {
+    color: #666;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:67]: same

File: src/app/pagination/pagination.component.css [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  .post-info {
+    margin-top: 15px;
+    font-size: 14px;
+    color: #555;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:73]: same

File: src/app/pagination/pagination.component.css [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  .tag {
+    display: inline-block;
+    background-color: #007bff;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:78]: same

File: src/app/pagination/pagination.component.css [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    font-size: 16px;
+    margin-right: 10px;
+    color: #333;
+  }
+
+  .pagination-controls select {
+    padding: 5px 10px;
+    border-radius: 5px;
+    border: 1px solid #ccc;
+    background-color: #fff;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:116]: same

File: src/app/pagination/pagination.component.css [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+    transition: background-color 0.3s, color 0.3s;
+  }
+
+  .pages:hover {
+    background-color: #007bff;
+    color: white;
+  }
+
+  .pages.active {
+    background-color: #007bff;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.css:180]: change all color with variables

File: src/app/pagination/pagination.component.html [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      <button *ngFor="let page of pages" 
+              class="pages" 
+              (click)="goToPage(page)" 
+              [ngClass]="{'active': page === currentPage}">
+        {{ page }}
+      </button>
+    </div>
+  </div>
+
+  <!-- Loader Container -->
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.html:43]: loader template & use them, functionality, we can pass the image source, text and more.

File: src/app/pagination/pagination.component.ts [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+  pageSize: number = 10;
+  currentPage: number = 1;
+  totalPages: number = 0;
+  pageSizes: number[] = [10, 20, 50, 100];
+  pages: number[] = [];
+  currentPagesIn: number[] = [];
+  loading: boolean = false;
+  private scrollSubscription!: Subscription;
+
+  constructor(private postService: PostService, private apiService: PostsService) {}
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:32]: Read about Dependency Injection

File: src/app/pagination/pagination.component.ts [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+    const scroll$ = fromEvent(this.postsContainer.nativeElement, 'scroll');
+    this.scrollSubscription = scroll$.pipe(
+      
+      debounceTime(100)
+      
+      // Debounce with a delay of 300ms
+    ).subscribe(() => this.onScroll());
+  }
+
+  ngOnDestroy(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:46]: bottom of the page

File: src/app/pagination/pagination.component.ts [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+  ): void {
+    if (this.loading) return;
+    this.loading = true;
+
+    const skip = prev ? loadedRecords : (this.currentPage - 1) * this.pageSize;
+    const limit = prev ? remainingRecord : this.pageSize;
+
+    this.postService.getPosts(limit, skip).subscribe(
+      (response) => {
+        const data = response.data;
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:74]: model, PostModel, responsibility _-> take response data, give data in my structure.

File: src/app/pagination/pagination.component.ts [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+        }, 0);
+      },
+      (error) => {
+        console.error('Error loading posts:', error);
+        this.loading = false;
+      }
+    );
+  }
+
+  onPageSizeChange(event: Event): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:105]: separated component, all implementation of the pagination should be there.

File: src/app/pagination/pagination.component.ts [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+      this.currentPage -= 1;
+      if (!this.currentPagesIn.includes(this.currentPage)) {
+        this.loadPosts(true);
+      } else {
+        this.adjustScroll();
+      }
+    }
+  }
+
+  nextPage(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:129]: move in pagination compopnent

File: src/app/pagination/pagination.component.ts [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+      this.currentPage += 1;
+      if (!this.currentPagesIn.includes(this.currentPage)) {
+        this.loadPosts(true);
+      } else {
+        this.adjustScroll();
+      }
+    }
+  }
+
+  goToPage(page: number): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:140]: move in pagination component

File: src/app/pagination/pagination.component.ts [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+  }
+
+  addPageToCurrentPages(page: number): void {
+    if (!this.currentPagesIn.includes(page)) {
+      this.currentPagesIn.push(page);
+      this.sortPages();
+    }
+  }
+
+  sortPages(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:163]: move in pagination component

File: src/app/pagination/pagination.component.ts [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+    this.currentPagesIn.sort((a, b) => a - b);
+  }
+
+  onScroll(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:167]: move in pagination component

File: src/app/pagination/pagination.component.ts [hunk 26]
```diff
@@ -0,0 +1,278 @@
...
+      !this.currentPagesIn.includes(this.currentPage + 1)
+    ) {
+      this.currentPage += 1;
+      this.addPageToCurrentPages(this.currentPage);
+      this.loadPosts(true);
+      this.adjustScroll();
+    }
+  }
+
+  adjustScroll(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:218]: move in pagination component

File: src/app/pagination/pagination.component.ts [hunk 27]
```diff
@@ -0,0 +1,278 @@
...
+    console.log(`Current Pages: ${this.currentPagesIn.length}`);
+    const pageHeight = container.scrollHeight / this.currentPagesIn.length;
+    console.log(`Page Height: ${pageHeight}`);
+    const index = this.getIndex(this.currentPagesIn, this.currentPage);
+    console.log(`Index: ${index}`);
+    console.log(`Required Height: ${pageHeight * index}`);
+    container.scrollTop = (pageHeight * index) + 15;
+  }
+
+  adjustPaginationScroll(): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:232]: move in pagination component

File: src/app/pagination/pagination.component.ts [hunk 28]
```diff
@@ -0,0 +1,278 @@
...
+    const paginationContainer = this.paginationContainer.nativeElement;
+    const pageWidth = paginationContainer.scrollWidth / this.totalPages;
+    const targetPosition = pageWidth * this.currentPage;
+    const leftScroll = targetPosition - paginationContainer.clientWidth;
+
+    paginationContainer.scrollLeft = leftScroll;
+  }
+
+  updatePagesOnPageSizeChange(newLimit: number, prevPageSize: number): void {
```
Comment (thesaty4) [inline src/app/pagination/pagination.component.ts:241]: move in pagination component

File: src/app/post.service.ts [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+import { HttpClient } from '@angular/common/http';
+import { Observable, of } from 'rxjs';
+import { tap } from 'rxjs/operators';
+import { CacheService } from './cache.service';
+import { APIResponse } from './services/api.service';
+import { ProductResponse } from './services/api-calling.service';
+
+
+@Injectable({
+  providedIn: 'root'
```
Comment (thesaty4) [inline src/app/post.service.ts:12]: not always root

File: src/app/post.service.ts [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+})
+export class PostService {
+  private apiUrl = 'https://dummyjson.com/posts';
```
Comment (thesaty4) [inline src/app/post.service.ts:15]: where is baseAPI ?

File: src/app/post.service.ts [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+import { tap } from 'rxjs/operators';
+import { CacheService } from './cache.service';
+import { APIResponse } from './services/api.service';
+import { ProductResponse } from './services/api-calling.service';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PostService {
```
Comment (thesaty4) [inline src/app/post.service.ts:14]: move it in the baseService

File: src/app/services/api-calling.service.ts [hunk 32]
```diff
@@ -0,0 +1,46 @@
+import { Injectable } from "@angular/core";
+import ApiService, {APIResponse} from "../services/api.service"
+
+export type PostResponse = {
```
Comment (thesaty4) [inline src/app/services/api-calling.service.ts:4]: types folder

File: src/app/services/api-calling.service.ts [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+export type ProductResponse = {
+  posts: PostResponse[];
+  total: number;
+  skip: number;
+  limit: number;
+};
+
+
+@Injectable({
+    providedIn: 'root', // This ensures the service is provided at the root level
```
Comment (thesaty4) [inline src/app/services/api-calling.service.ts:26]: saa

File: src/app/services/api-calling.service.ts [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+  })
+
+  export class PostsService extends ApiService {
```
Comment (thesaty4) [inline src/app/services/api-calling.service.ts:29]: file name

File: src/app/services/api-calling.service.ts [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+      return this.get<ProductResponse>(`posts?limit=${limit}&skip=${skip}`)
+        .then(response => {
+          console.log('API response:', response); // Log the response to inspect its structure
+  
+          
+            return response; // Return the response directly if it matches the expected structure
+          
+          // If the response is not in expected format
+        })
+        .catch(error => {
```
Comment (thesaty4) [inline src/app/services/api-calling.service.ts:40]: error handling in the getPost calling

File: src/app/services/api.service.ts [hunk 36]
```diff
@@ -0,0 +1,140 @@
+import axios, { AxiosInstance, AxiosResponse, AxiosError, InternalAxiosRequestConfig } from 'axios';
+
+export default class RequestService {
+  private axiosInstance: AxiosInstance;
+
+  constructor() {
+    // Create an Axios instance
+    this.axiosInstance = axios.create({
```
Comment (thesaty4) [inline src/app/services/api.service.ts:8]: use httpRequest

File: src/app/services/api.service.ts [hunk 37]
```diff
@@ -0,0 +1,140 @@
...
+      baseURL: 'https://dummyjson.com/',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      timeout: 3000,
+    });
+
+    // Request interceptor
+    this.axiosInstance.interceptors.request.use(
```
Comment (thesaty4) [inline src/app/services/api.service.ts:17]: remove

File: src/app/services/api.service.ts [hunk 38]
```diff
@@ -0,0 +1,140 @@
...
+        return config;
+      },
+      (error: AxiosError) => {
+        console.error('Request error:', error);
+        return Promise.reject(error);
+      }
+    );
+
+    // Response interceptor
+    this.axiosInstance.interceptors.response.use(
```
Comment (thesaty4) [inline src/app/services/api.service.ts:31]: remove

File: src/app/services/api.service.ts [hunk 39]
```diff
@@ -0,0 +1,140 @@
...
+    header?: Record<string, any>,
+  ): Promise<APIResponse<T>> {
+    try {
+      const response: AxiosResponse<APIResponse<T>> = await this.axiosInstance.request({
+        method,
+        url: endpoint,
+        headers: {
+          ...header,
+        },
+        data: (header?.['Content-Type'] === 'multipart/form-data'
```
Comment (thesaty4) [inline src/app/services/api.service.ts:60]: copy past ?

File: src/app/services/api.service.ts [hunk 40]
```diff
@@ -0,0 +1,140 @@
...
+          throw new Error(errorMessage);
+        } else {
+          // If no specific error message from the server, throw a generic error
+          throw new Error(error?.message);
+        }
+      }
+    }
+  }
+
+  protected async get<T>(
```
Comment (thesaty4) [inline src/app/services/api.service.ts:96]: deep dive

File: src/app/services/api.service.ts [hunk 41]
```diff
@@ -0,0 +1,140 @@
...
+    data?: T,
+    header?: Record<string, any>,
+  ): Promise<APIResponse<T>> {
+    return this.makeRequest('post', endpoint, data, header);
+  }
+
+  protected async delete<T>(
+    endpoint: string,
+    data?: T,
+  ): Promise<APIResponse<T>> {
```
Comment (thesaty4) [inline src/app/services/api.service.ts:115]: Observable

File: src/app/services/api.service.ts [hunk 42]
```diff
@@ -0,0 +1,140 @@
...
+    return this.makeRequest('delete', endpoint, data);
+  }
+
+  protected async put<T>(
+    endpoint: string,
+    data?: any,
+    header?: Record<string, any>,
+  ): Promise<APIResponse<T>> {
```
Comment (thesaty4) [inline src/app/services/api.service.ts:123]: change observable

**Review by thesaty4 — CHANGES_REQUESTED** [review]
fix review points

File: src/app/api.interceptor.ts (hunk 1 above)
Comment (ADDY69O) [reply src/app/api.interceptor.ts:5]: Done

**Pull request merged**